File: backend-haskell/src/BackendHaskell/Auth/UserResolver.hs
-- | REST/gRPCの両トランスポートが共有するuser_id解決ロジック(認証ロジックを複製しない設計)。
-- backend(Go)のresolveUserID・backend-java/backend-kotlin/backend-python/backend-elixirの
-- UserResolverと同じ分岐:
--     - ローカル(HMAC/RSA)発行のJWT: subは内部user_idそのもの -> usersをidで検索
--     - Keycloak発行のJWT: subはkeycloak_sub -> user_keycloaks経由で検索
-- どちらも見つからなければuser_not_provisioned
module BackendHaskell.Auth.UserResolver
  ( resolveUserId
  ) where

import Data.Pool (Pool)
import Data.Text (Text)
import qualified Data.Text as T
import Text.Read (readMaybe)

import BackendHaskell.Auth.Claims (Claims (..))
import BackendHaskell.Auth.Dispatcher (Dispatcher, dispatchVerify, isLocalIssuer)
import BackendHaskell.Domain.Errors (TaskError (..), TaskErrorKind (..))
import BackendHaskell.Domain.Models (User (..))
import Database.MySQL.Base (MySQLConn)
import qualified BackendHaskell.Repository.TaskRepository as Repo

resolveUserId :: Pool MySQLConn -> Dispatcher -> Maybe Text -> IO (Either TaskError Int)
resolveUserId pool dispatcher maybeAuthHeader =
  case maybeAuthHeader >>= stripBearer of
    Nothing -> pure (Left unauthorized)
    Just token -> do
      verifyResult <- dispatchVerify dispatcher token
      case verifyResult of
        Left _ -> pure (Left unauthorized)
        Right claims -> resolveFromClaims pool claims

stripBearer :: Text -> Maybe Text
stripBearer header =
  let prefix = "Bearer "
   in if T.isPrefixOf prefix header && T.length header > T.length prefix
        then Just (T.drop (T.length prefix) header)
        else Nothing

resolveFromClaims :: Pool MySQLConn -> Claims -> IO (Either TaskError Int)
resolveFromClaims pool claims
  | isLocalIssuer (claimsIss claims) = case readMaybe (T.unpack (claimsSub claims)) :: Maybe Int of
      Nothing -> pure (Left userNotProvisioned)
      Just uid -> do
        found <- Repo.findUserById pool uid
        pure (maybe (Left userNotProvisioned) (Right . userId) found)
  | otherwise = do
      found <- Repo.findUserByKeycloakSub pool (claimsSub claims)
      pure (maybe (Left userNotProvisioned) (Right . userId) found)

unauthorized :: TaskError
unauthorized = TaskError Unauthorized "unauthorized"

userNotProvisioned :: TaskError
userNotProvisioned = TaskError UserNotProvisioned "user_not_provisioned"

File: backend-python/app/auth/user_resolver.py
from __future__ import annotations

import logging

from app.auth.claims import VerifyException
from app.auth.dispatcher import Dispatcher, is_local_issuer
from app.domain.errors import TaskError
from app.repository.task_repository import TaskRepository

log = logging.getLogger(__name__)


class UserResolver:
    """REST/gRPCの両トランスポートが共有するuser_id解決ロジック(認証ロジックを複製しない設計)。
    backend(Go)のresolveUserID・backend-rust/backend-java/backend-kotlinのUserResolverと同じ分岐:
        - ローカル(HMAC/RSA)発行のJWT: subは内部user_idそのもの -> usersをidで検索
        - Keycloak発行のJWT: subはkeycloak_sub -> user_keycloaks経由で検索
    どちらも見つからなければUserNotProvisioned
    """

    def __init__(self, dispatcher: Dispatcher, repository: TaskRepository) -> None:
        self._dispatcher = dispatcher
        self._repository = repository

    async def resolve(self, authorization_header: str | None) -> int:
        if authorization_header is None or not authorization_header.startswith("Bearer "):
            raise TaskError.unauthorized()
        token = authorization_header[len("Bearer ") :]

        try:
            claims = await self._dispatcher.verify(token)
        except VerifyException as exc:
            raise TaskError.unauthorized() from exc

        if is_local_issuer(claims.iss):
            try:
                user_id = int(claims.sub)
            except ValueError as exc:
                raise TaskError.user_not_provisioned() from exc
            user = await self._repository.find_user_by_id(user_id)
            if user is None:
                raise TaskError.user_not_provisioned()
            log.debug("resolved user_id=%d via local issuer=%s", user.id, claims.iss)
            return user.id

        user = await self._repository.find_user_by_keycloak_sub(claims.sub)
        if user is None:
            raise TaskError.user_not_provisioned()
        log.debug("resolved user_id=%d via keycloak_sub=%s", user.id, claims.sub)
        return user.id
